### PR TITLE
fipsccl: Add interfaces to expose FIPS-readiness status

### LINKED ISSUE #114344: cliccl: Add interfaces to expose FIPS status
**Is your feature request related to a problem? Please describe.**

Our [current FIPS configuration process](https://www.cockroachlabs.com/docs/stable/fips) includes many steps, and the final verification step at the end is a very weak check (it confirms that you are using a FIPS binary but not that all the other parts of the process were successful. The FIPS binary will still fall back to non-FIPS mode unless everything is correct).

This is particularly important for 23.2 because this release coincides with our move from `ubi8` to `ubi9` which entails a move from OpenSSL 1.1 to 3.0. FIPS builds must be deployed with the same OpenSSL version that was used to build them in order to use FIPS mode (failure to do so currently means a silent fallback to non-FIPS mode).

**Describe the solution you'd like**

One or more interfaces to verify the use of OpenSSL in as comprehensive a way as possible. Specifically,
- A command-line flag `--enterprise-require-fips`. If this flag is given, the `cockroach` command will exit with an error if any of the preconditions for FIPS compliance are not met (this flag would be present in all builds, but in builds that are not fips-ready it will always fail). This will be a global flag and can be given on any `cockroach` subcommand (including for example `cockroach sql`) as these commands may also use crypto features that require FIPS in some environments. 
- A SQL function `crdb_internal.require_fips()` which returns `true` if the gateway node's FIPS configuration is valid, and raises an error otherwise. This is redundant with the CLI flag but it gives customers an option to periodically re-check if desired.

These are both enterprise features (It would still be possible to build a FIPS-ready build without enterprise features, but enforcing compliance would be in enterprise). The SQL function performs a license check as usual. The `--enterprise-require-fips` flag has the same status as `--enterprise-encryption`: it does not have a programmatic license check because we must be able to decrypt our on-disk data before we can see the license status.

**Describe alternatives you've considered**

The `--enterprise-require-fips` functionality could be performed by default in a FIPS-ready build. This would seem to remove a step but in practice would just replace "use the `--enterprise-require-fips` flag" with "run `cockroach version` to make sure you have the correct binary".  

**Additional context**

These features should at least be backported to 23.2 given the risk of issues with the OpenSSL 3.0 migration happening concurrently. 

Jira issue: CRDB-33472

### PR DESCRIPTION
**Pull request by bdarnell**
- CLI tool `cockroach debug enterprise-check-fips` for detailed diagnostics
- Global CLI flag `--enterprise-require-fips-ready` to abort if FIPS checks fail
- SQL function `crdb_internal.fips_ready()` to check at runtime

Closes #114344

Even though FIPS is security-related (and I added a `securityccl` package to be owned by @cockroachdb/prodsec), this PR is really more build wrangling than anything significant from a security perspective.

### CODE REVIEW

**Review by rsevinsky-cr — APPROVED** [review]
> Reviewed 1 of 1 files at r1, 9 of 9 files at r2, 3 of 3 files at r3, 23 of 23 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/114709)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @bdarnell)

**Review by rickystewart — APPROVED** [review]

File: pkg/ccl/securityccl/fipsccl/build_boring.go [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+import "C"
+
+import (
+	"crypto/boring"
+	"fmt"
+)
+
+// IsBoringBuild returns true if this binary was built with the boringcrypto
+// build tag, which is a prerequisite for FIPS-ready mode.
+func IsBoringBuild() bool {
```
[the remark this answers is not in the document]
Comment (bdarnell) [reply pkg/ccl/securityccl/fipsccl/build_boring.go]: OK, renamed to `IsCompileTimeFIPSReady`.

**Review by rsevinsky-cr — APPROVED** [review]
> Reviewed 29 of 29 files at r5, 3 of 3 files at r6, 23 of 23 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/114709)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rickystewart)

**Pull request merged**